crates/ironrdp-pdu/src/utils.rs

```diff
@@ -26,6 +27,98 @@ pub(crate) enum CharacterSet {
     Unicode = 2,
 }
 
+pub(crate) fn read_string_from_cursor(
+    cursor: &mut ReadCursor<'_>,
+    character_set: CharacterSet,
+    read_null_terminator: bool,
+) -> PduResult<String> {
+    let size = if character_set == CharacterSet::Unicode {
+        let code_points = if read_null_terminator {
+            // Find null or read all if null is not found
+            cursor
+                .remaining()
+                .chunks_exact(2)
+                .position(|chunk| chunk[0] == 0 && chunk[1] == 0)
+                .map(|codepoints| codepoints + 1) // Read null codepoint
+                .unwrap_or(cursor.len() / 2)
+        } else {
+            // UTF16 uses 2 bytes per code point, so we need to read an even number of bytes
+            cursor.len() / 2
```

Comment:
I know this is something done in the current codebase, that dates back to when IronRDP was initially written, but that’s not true. I opened an issue on this: https://github.com/Devolutions/IronRDP/issues/130

Reply:
No, it's a different calculation -- not the size of the UTF16 string, but the closest valid buffer size which could be used for a decoding attempt.

I think I just used the wrong naming, which gave this confusion -- I think I wanted to write `code units` instead of `code points`. Code units in UTF16 are always 2 bytes in size.

Reply:
Oh I see. Fair enough!